### PR TITLE
Convert Makefile to CMakeLists.txt

### PR DESCRIPTION
**Pull request by cocoa-xu**
Still WIP, but it's close

### CODE REVIEW

**Comment by polvalente** [comment]
Is there a specific motivation for this change?

File: torchx/CMakeLists.txt [hunk 1]
```diff
@@ -6,39 +6,91 @@ if(NOT DEFINED CMAKE_CUDA_ARCHITECTURES)
...
+    set(ERTS_INCLUDE_DIR "${ERTS_INCLUDE_DIR}")
+else()
+    set(ERTS_INCLUDE_DIR_ONE_LINER "erl -noshell -eval \"io:format('~ts/erts-~ts/include/', [code:root_dir(), erlang:system_info(version)]), halt().\"")
+    if(WIN32)
+        execute_process(COMMAND powershell -command "${ERTS_INCLUDE_DIR_ONE_LINER}" OUTPUT_VARIABLE ERTS_INCLUDE_DIR)
+    else()
+        execute_process(COMMAND bash -c "${ERTS_INCLUDE_DIR_ONE_LINER}" OUTPUT_VARIABLE ERTS_INCLUDE_DIR)
+    endif()
+    set(ERTS_INCLUDE_DIR "${ERTS_INCLUDE_DIR}")
+endif()
```
Comment (josevalim) [inline torchx/CMakeLists.txt:37]: We can still compute this on the Elixir side and pass it as `env` in the `cmd!` function. WDYT?

File: torchx/CMakeLists.txt [hunk 2]
```diff
@@ -6,39 +6,91 @@ if(NOT DEFINED CMAKE_CUDA_ARCHITECTURES)
...
+message(STATUS "C_SRC: ${C_SRC}")
+message(STATUS "MIX_APP_PATH: $ENV{MIX_APP_PATH}")
+message(STATUS "PRIV_DIR: ${PRIV_DIR}")
+message(STATUS "ERTS_INCLUDE_DIR: ${ERTS_INCLUDE_DIR}")
+message(STATUS "LIBTORCH_DIR: $ENV{LIBTORCH_DIR}")
+message(STATUS "LIBTORCH_BASE: $ENV{LIBTORCH_BASE}")
+message(STATUS "MIX_BUILD_EMBEDDED $ENV{MIX_BUILD_EMBEDDED}")
+message(STATUS "LIBTORCH_LINK $ENV{LIBTORCH_LINK}")
+
+if((NOT DEFINED ENV{LIBTORCH_DIR}) OR (NOT IS_DIRECTORY $ENV{LIBTORCH_DIR}))
```
Comment (josevalim) [inline torchx/CMakeLists.txt:52]: We can also move this check to our `cmake` function? :)

File: torchx/mix.exs [hunk 3]
```diff
@@ -252,4 +238,54 @@ defmodule Torchx.MixProject do
...
+    opts = [
+      into: IO.stream(:stdio, :line),
+      stderr_to_stdout: true
+    ]
+
+    case System.cmd(exec, args, opts) do
+      {_, 0} -> :ok
+      {_, status} -> Mix.raise("cmake failed with status #{status}")
+    end
+  end
```
Comment (josevalim) [inline torchx/mix.exs:290]:
```suggestion
    env = %{
      "LIBTORCH_DIR" => libtorch_config.dir,
      "LIBTORCH_BASE" => libtorch_config.base,
      "MIX_BUILD_EMBEDDED" => "#{Mix.Project.config()[:build_embedded]}",
      "LIBTORCH_LINK" => "#{libtorch_link_path}/lib",
      "MIX_APP_PATH" => Mix.Project.app_path()
    }

    cmd!(cmake, ["-S", ".", "-B", cmake_build_dir], env)
    cmd!(cmake, ["--build", cmake_build_dir, "--config", cmake_build_type], env)
    cmd!(cmake, ["--install", cmake_build_dir, "--config", cmake_build_type], env)

    {:ok, []}
  end

  defp cmd!(exec, args, env) do
    opts = [
      into: IO.stream(:stdio, :line),
      stderr_to_stdout: true,
      env: env
    ]

    case System.cmd(exec, args, opts) do
      {_, 0} -> :ok
      {_, status} -> Mix.raise("cmake failed with status #{status}")
    end
  end
```

File: torchx/CMakeLists.txt (hunk 1 above)
Comment (cocoa-xu) [reply torchx/CMakeLists.txt:37]: Sure! I'll copy some relevant code from `elixir_make` to do it :)

File: torchx/CMakeLists.txt (hunk 2 above)
Comment (cocoa-xu) [reply torchx/CMakeLists.txt]: Will do!

**Comment by josevalim** [comment]
Beautiful, this is a massive contribution to the user experience :green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

**Comment by cocoa-xu** [comment]
I got this error in the Linux CI. Not quite sure if this error had occurred in other CI runs before, but it seems to be not relevant to this PR. WDYT?

```
  1) test batched_run 2=>shutdown=>2 (queued) (Nx.ServingTest)
Error:      test/nx/serving_test.exs:617
     Assertion failed, no matching message after 1000ms
     Showing 2 of 2 messages in the mailbox
     code: assert_receive {:DOWN, ref, _, _, {:noproc, {Nx.Serving, :local_batched_run, [_, _]}}} when ref in [ref1, ref2]
     mailbox:
       pattern: {:DOWN, ref, _, _, {:noproc, {Nx.Serving, :local_batched_run, [_, _]}}} when ref in [ref1, ref2]
       value:   {#Reference<0.52196280.564658178.91856>, :ok}

       pattern: {:DOWN, ref, _, _, {:noproc, {Nx.Serving, :local_batched_run, [_, _]}}} when ref in [ref1, ref2]
       value:   {:execute, 0, #PID<0.3011.0>}
     stacktrace:
       test/nx/serving_test.exs:640: (test)

.........................
Finished in 8.1 seconds (8.1s async, 0.00s sync)
1301 doctests, 1101 tests, 1 failure

Randomized with seed 154005
```

**Comment by josevalim** [comment]
It is a race condition in the test suite, I will look into it. It is not related, so no worries!

**Pull request merged**